File: public/docs/_examples/hierarchical-dependency-injection/ts/app/hero-editor.component.ts
// #docregion
import { Component, EventEmitter, Input, Output } from '@angular/core';

import { RestoreService } from './restore.service';
import { Hero } from './hero';

@Component({
  selector: 'hero-editor',
  // #docregion providers
  providers: [RestoreService],
  // #enddocregion providers
  template: `
    <div>
      <span>Name:</span>
      <input [(ngModel)]="hero.name"/>
      <div>
        <button (click)="onSaved()">save</button>
        <button (click)="onCanceled()">cancel</button>
      </div>
    </div>`
})

export class HeroEditorComponent {
  @Output() canceled = new EventEmitter<Hero>();
  @Output() saved = new EventEmitter<Hero>();

  constructor(private restoreService: RestoreService<Hero>) {}

  @Input()
  set hero (hero: Hero) {
    this.restoreService.setItem(hero);
  }

  get hero () {
    return this.restoreService.getItem();
  }

  onSaved () {
    this.saved.emit(this.restoreService.getItem());
  }

  onCanceled () {
    this.hero = this.restoreService.restoreItem();
    this.canceled.emit(this.hero);
  }
}
// #enddocregion

File: public/docs/_examples/hierarchical-dependency-injection/dart/lib/hero_editor_component.dart
// #docregion
import 'package:angular2/core.dart';

import 'hero.dart';
import 'restore_service.dart';

@Component(
    selector: 'hero-editor',
    // #docregion providers
    providers: const [RestoreService],
    // #enddocregion providers
    template: '''
  <div>
    <span>Name:</span>
    <input [(ngModel)]="hero.name"/>
    <div>
      <button (click)="onSaved()">save</button>
      <button (click)="onCanceled()">cancel</button>
    </div>
  </div>
  ''')
class HeroEditorComponent {
  @Output() final EventEmitter canceled = new EventEmitter();
  @Output() final EventEmitter saved = new EventEmitter();

  RestoreService<Hero> _restoreService;

  HeroEditorComponent(this._restoreService);

  @Input()
  set hero(Hero hero) {
    _restoreService.setItem(hero);
  }

  Hero get hero {
    return _restoreService.getItem();
  }

  onSaved() {
    saved.add(_restoreService.getItem());
  }

  onCanceled() {
    hero = _restoreService.restoreItem();
    canceled.add(hero);
  }
}
// #enddocregion
